### PR TITLE
Adds the ability to download a borg's logs using SiliConnect.

### PR DESCRIPTION
**Pull request by zxaber**
## About The Pull Request
Adds a function to Siliconnect in which you can tap a borg using a mobile device that is currently running the software, and it will download the borg's logs. These logs list events such as being locked or unlocked, taken offline and being restored, and the process of being emagged. Borg logs will now list the emag user's name as a new user.

Downloading the logs is done by right-clicking the borg while the software is open and active. This will "tap" the borg with the device, initiating the log transfer automatically. The transfer will take eight seconds to complete, requiring you and the borg to stay adjacent the entire time, and the borg will get a large red text alert about the upload when it starts. The logs are also not stored permanently on the device, and will be lost if the app is closed.

This PR also introduces the `tap()` proc for modular computer programs, and adds functionality the the computers to call it when right-clicking with the tablet as a tool. Should the app use the tap in a meaningful way (such as starting a borg log download), it will return TRUE, and the computer will end the secondary attack chain.

Currently, this requires using a tablet or laptop with Siliconnect, as you cannot tap with a console. Someday I hope to add an additional hardware option for consoles in the form of a wireless hand scanner to replicate tapping.

## Why It's Good For The Game
Adds a neat way to "diagnose" a borg acting oddly, assuming you have a way to keep them still. Allows one to view if the borg has been emagged, by whom, what SiliConnect messages the borg has received, the number of law changes that have been made, as well as some other (somewhat fluff) information relating to taking damage and getting upgrades.

## Changelog
:cl:
add: You can now download a cyborg's internal logs by right-clicking them with a mobile device running SiliConnect. Take a look if one is acting a bit strange, you might find something interesting.
add: Borg integrity (health) is now roughly shown in SiliConnect, under "Condition".
/:cl:

To Do:

- [x] Fix the to_chat not being sent to the borg when a log transfer breaks due to distance
- [x] Add some sort of text feedback when tapping an atom with a device, or viewing one being tapped.
- [x] ~~Possibly restrict syndicate borgs from having their logs viewed by SiliConnect (and likewise with station borgs and Roboverlord)~~ Changed my mind on this, will implement later if we find it's needed.
- [x] Change the log area of the window to expand naturally rather than use a fixed height

image

I'm not super deadset on the emag user's name being listed. It seems like a neat function unique to having a tablet running Siliconnect (as most of the other info is already available through other means), and one can always pop the brain out for a re-borg and demand the name anyway.

### CODE REVIEW

**Comment by Dorsisdwarf** [comment]
The log should probbably log the ID of the emagger, rather than their secret real name no?

**Comment by XDTM** [comment]
It appears to use the public name, so someone in a gasmask would appear as unknown.
Although considering that the emagger's real name is written into the emag laws automatically, there is a case to be made for using that same name in the borg logs.

**Comment by zxaber** [comment]
I had meant to use their real name, yeah. While the ability to hide your tracks using a mask would be neat, the name in the logs is supposed to mirror the name being used for the lawset. If the emag tool had players manually enter a name, I'd use that, but there's probably not a good way to handle such in the case of emagging a borg.

**Comment by zxaber** [comment]
- Now uses real_name for the emag log.
- All uses of `Flex` in the tgui code have been replaced with `Stack`
- <details><summary>
  Stack allows for auto-expanded sections, so the log box always follows the window size now
  </summary>

  image
  </details>
- <details><summary>
  Also, since I'm working on this app already, I added the borg health to the app, something I had meant to do before.
  </summary>
  "Functionality Impaired" means the borg has lost at least one module slot, "Functionality Disrupted" means all three are gone, and "Hard Fault" means the borg is out of health.

  image
  </details>

File: tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js [hunk 1]
```diff
@@ -30,64 +50,134 @@ export const NtosCyborgRemoteMonitorContent = (props, context) => {
...
+            selected={tab_main === 2}
+            onClick={() => setTab_main(2)}>
+            Stored Log File
+          </Tabs.Tab>
+        </Tabs>
+      </Stack.Item>
+      {tab_main === 1 && (
+        <>
+          {!card && (
+            <NoticeBox>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js:75]: Is missing a stack item around it.

File: tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js [hunk 2]
```diff
@@ -30,64 +50,134 @@ export const NtosCyborgRemoteMonitorContent = (props, context) => {
...
+              Certain features require an ID card login.
+            </NoticeBox>
+          )}
+          <Stack.Item grow={1}>
+            <Section fill scrollable>
+              {cyborgs.map(cyborg => {
+                return (
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js:82]: compactify into return-less `{cyborgs.map(cyborg => (`, reduce final indentation by one.

**Comment by stylemistake** [comment]
i'll merge once you do that, also check the commit above, you were importing a fragment from the wrong place, and it wasn't needed at all.

File: tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js (hunk 1 above)
Comment (zxaber) [reply tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js:75]: Is a `<Stack.Item>` here necessary? It works without it, and putting one in breaks the UI.

File: tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js (hunk 2 above)
Comment (zxaber) [reply tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js]: Done

File: tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js (hunk 1 above)
Comment (stylemistake) [reply tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js:75]: Yes, breaks how

Comment (zxaber) [reply tgui/packages/tgui/interfaces/NtosCyborgRemoteMonitor.js]: Nevermind, I misunderstood where you wanted the stacks.

This is done now.

**Pull request merged**